### PR TITLE
Contract For Swapping Tokens

### PR DESCRIPTION
**Pull request by efecarranza**
# Changelog

Add contract for swapping tokens.
Add tests for swapping contract.

### CODE REVIEW

File: src/AaveV2CollectorContractConsolidation.sol [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+        uint48 premium;
+        address oracle;
+        bool ethFeedOnly;
+    }
+
+    mapping(address => Asset) public assets;
+
+    constructor() {
+        assets[ARAI] = Asset(6740239e16, 100, 0x483d36F6a1d063d580c7a24F9A42B346f3a69fbb, false); // Custom Feed
+        assets[AAMPL] = Asset(15891248e7, 300, 0xe20CA8D7546932360e37E9D72c1a47334af57706, false); // Monitored Feed
```
Comment (0xrajath) [inline src/AaveV2CollectorContractConsolidation.sol]: What's a custom and monitored feed?

File: src/AaveV2CollectorContractConsolidation.sol [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+        assets[ATUSD] = Asset(608004e16, 75, 0xec746eCF986E2927Abd291a2A1716c940100f8Ba, false);
+        assets[AMANA] = Asset(1622740e16, 200, 0x56a4857acbcfe3a66965c251628B1c9f1c408C19, false);
+        assets[MANA] = Asset(33110e16, 200, 0x56a4857acbcfe3a66965c251628B1c9f1c408C19, false);
+        assets[ADPI] = Asset(1359e16, 300, 0xD2A593BF7594aCE1faD597adb697b5645d5edDB2, false); 
+        assets[ABUSD] = Asset(364085e16, 75, 0x833D8Eb16D306ed1FbB5D7A2E019e106B960965A, false);
+        assets[BUSD] = Asset(3399064e14, 75, 0x833D8Eb16D306ed1FbB5D7A2E019e106B960965A, false);
+        assets[ZRX] = Asset(10719e16, 300, 0x2885d15b8Af22648b98B122b22FDF4D2a56c6023, false);
+        assets[AZRX] = Asset(877140e16, 300, 0x2885d15b8Af22648b98B122b22FDF4D2a56c6023, false);
+        assets[AUST] = Asset(89239797e4, 200, address(0), false); // NO FEED
+        assets[ARENFIL] = Asset(41067e16, 300, address(0), false); // NO FEED
```
Comment (0xrajath) [inline src/AaveV2CollectorContractConsolidation.sol]: What're you doing in the case of No feed?

File: src/AaveV2CollectorContractConsolidation.sol [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+    /// @notice Returns amount of USDC to be spent to swap for token
+    /// @param _token the address of the token to swap
+    /// @param _amountOut the amount of token wanted
+    /// @param _decimals decimals of ERC20 token
+    /// return amountInWithDiscount the amount of USDC used minus premium incentive
+    /// @dev User check this function before calling swap() to see the amount of USDC required
+    function getAmountIn(
+        address _token,
+        uint256 _amountOut,
+        uint256 _decimals
```
Comment (0xrajath) [inline src/AaveV2CollectorContractConsolidation.sol]: Do you need to pass the decimals in as a parameter separately? You can get it from the `_token` itself right?

File: src/AaveV2CollectorContractConsolidation.sol [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+        if (_amountOut == 0) revert OnlyNonZeroAmount();
+
+        uint256 amountIn = getAmountIn(_token, _amountOut, ERC20(_token).decimals());
+
+        uint256 quantity = assets[_token].quantity;
+        uint256 sendAmount = _amountOut == type(uint256).max ? quantity : _amountOut;
+
+        unchecked {
+            assets[_token].quantity = quantity - sendAmount;
+        }
```
Comment (0xrajath) [inline src/AaveV2CollectorContractConsolidation.sol]: Do you have a test case to check if this won't underflow?

Comment (0xrajath) [reply src/AaveV2CollectorContractConsolidation.sol]: How much gas are you saving by doing this?

File: src/AaveV2CollectorContractConsolidation.sol [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+        if (asset.oracle == address(0)) revert UnsupportedToken();
+
+        if (_amountOut == type(uint256).max) {
+            _amountOut = asset.quantity;
+        } else if (_amountOut > asset.quantity) {
+            revert NotEnoughTokens();
+        }
+
+        (uint256 oraclePrice, uint8 oracleDecimals) = getOraclePrice(asset.oracle);
+        unchecked {
```
Comment (0xrajath) [inline src/AaveV2CollectorContractConsolidation.sol]: Do you have test cases to check that none of the statements in this scope of unchecked will overflow/underflow?

Comment (0xrajath) [reply src/AaveV2CollectorContractConsolidation.sol]: How much gas are you saving by doing this?

File: src/AaveV2CollectorContractConsolidation.sol [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+        uint256 _decimals
+    ) public view returns (uint256 amountIn) {
+        Asset memory asset = assets[_token];
+        if (asset.oracle == address(0)) revert UnsupportedToken();
+
+        if (_amountOut == type(uint256).max) {
+            _amountOut = asset.quantity;
+        } else if (_amountOut > asset.quantity) {
+            revert NotEnoughTokens();
+        }
```
Comment (0xrajath) [inline src/AaveV2CollectorContractConsolidation.sol]: I would do this code block as 

```
if (asset.oracle == address(0)) revert UnsupportedToken();
if (_amountOut > asset.quantity) revert NotEnoughTokens();

if (_amountOut == type(uint256).max) {
       _amountOut = asset.quantity;
}
```

This has all the checks before any other effects happen and follows the checks-effects-interaction security pattern.

File: src/AaveV2CollectorContractConsolidation.sol [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+
+            if (asset.ethFeedOnly) {
+                (uint256 ethUsdPrice, uint8 ethUsdDecimals) = getOraclePrice(ETH_USD_FEED);
+                oraclePrice *= ethUsdPrice;
+                exponent += ethUsdDecimals;
+            }
+
+            // Basis points arbitrage incentive
+            amountIn = ((_amountOut * oraclePrice / 10**exponent) // Amount in before discount 
+                * (10000 - asset.premium)) / 10000;
```
Comment (0xrajath) [inline src/AaveV2CollectorContractConsolidation.sol]: Math here seems to be wrong. If you're given a premium on that asset, then the discount on USD will be 
`* 10000 / (10000 + asset.premium)`

File: src/AaveV2CollectorContractConsolidation.sol (hunk 2 above)
Comment (0xrajath) [reply src/AaveV2CollectorContractConsolidation.sol]: Just saw that you're handling the case of address(0), in `getAmountIn()`. Why even have these assets here in the first place if they're not going to be supported?

File: src/test/AaveV2CollectorContractConsolidation.t.sol [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+        allTokens[17] = collectorContract.ARENFIL();
+    }
+
+    function testSwapZeroAmount() public {
+        address frax = collectorContract.FRAX();
+        vm.expectRevert(AaveV2CollectorContractConsolidation.OnlyNonZeroAmount.selector);
+        collectorContract.swap(frax, 0);
+    }
+
+    function testSwapAllAAmpl() public {
```
Comment (0xrajath) [inline src/test/AaveV2CollectorContractConsolidation.t.sol]: You should be able to collapse the test cases for each of these individual tokens into one test by looping through the finite list of assets within the test case. That probably makes more sense than to have 20 or so individual tests (i.e one for each asset)

File: src/test/AaveV2CollectorContractConsolidation.t.sol [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+pragma solidity 0.8.17;
+
+import "@forge-std/Test.sol";
+import "@forge-std/console.sol";
+import {AaveV2Ethereum} from "@aave-address-book/AaveV2Ethereum.sol";
+import {AaveV2CollectorContractConsolidation} from "../AaveV2CollectorContractConsolidation.sol";
+import {AggregatorV3Interface} from "../external/AggregatorV3Interface.sol";
+import {ERC20} from "@openzeppelin/token/ERC20/ERC20.sol";
+
+contract AaveV2CollectorContractConsolidationTest is Test {
```
Comment (0xrajath) [inline src/test/AaveV2CollectorContractConsolidation.t.sol]: Also have a test for some asset that's not initialized through the constructor. (i.e it should not work when you pass in that random token address)

File: src/AaveV2CollectorContractConsolidation.sol (hunk 1 above)
Comment (efecarranza) [reply src/AaveV2CollectorContractConsolidation.sol]: I was just making a note to sanity check these feeds: https://docs.chain.link/data-feeds/price-feeds/addresses/

File: src/AaveV2CollectorContractConsolidation.sol (hunk 2 above)
Comment (efecarranza) [reply src/AaveV2CollectorContractConsolidation.sol]: I'm actually waiting on BGD to see which Oracle to use, we need to support these tokens.

File: src/AaveV2CollectorContractConsolidation.sol (hunk 3 above)
Comment (efecarranza) [reply src/AaveV2CollectorContractConsolidation.sol]: Updated this, ended up caching the decimals.

File: src/AaveV2CollectorContractConsolidation.sol (hunk 6 above)
Comment (efecarranza) [reply src/AaveV2CollectorContractConsolidation.sol]: The thing with this one is that `_amountOut` if it's type(uint256).max, is going to be greater than quantity and it'll revert.

File: src/AaveV2CollectorContractConsolidation.sol (hunk 7 above)
Comment (efecarranza) [reply src/AaveV2CollectorContractConsolidation.sol]: Since I'm doing `amountIn` vs. `amountOut`, I think this is right. In terms of USDC, it'd be cheaper for the person sending USDC than the market value. So say, $1,000 of BUSD which should be 1:1, would require say, 997.5 USDC.

File: src/test/AaveV2CollectorContractConsolidation.t.sol (hunk 8 above)
Comment (efecarranza) [reply src/test/AaveV2CollectorContractConsolidation.t.sol]: Yeah, I'm sanity checking everything and getting exact values and debugging each test individually before collapsing into one.

File: src/test/AaveV2CollectorContractConsolidation.t.sol (hunk 9 above)
Comment (efecarranza) [reply src/test/AaveV2CollectorContractConsolidation.t.sol]: Will do.

File: src/AaveV2CollectorContractConsolidation.sol (hunk 4 above)
Comment (efecarranza) [reply src/AaveV2CollectorContractConsolidation.sol]: I can add a test case but it should not overflow since the sendAmount is being checked elsewhere against the quantity already.

It saves 70 gas total. Might not be worth it for this contract as we don't expect that many calls, but I feel ok about this one. I can go either way because the number of calls will be few.

File: src/AMMWithdrawer.sol [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+import {SafeERC20} from "@openzeppelin/token/ERC20/utils/SafeERC20.sol";
+
+contract AMMWithdrawer {
+    using SafeERC20 for IERC20;
+
+    address public constant AMMDAI = 0x79bE75FFC64DD58e66787E4Eae470c8a1FD08ba4;
+    address public constant AMMUSDC = 0xd24946147829DEaA935bE2aD85A3291dbf109c80;
+    address public constant AMMUSDT = 0x17a79792Fe6fE5C95dFE95Fe3fCEE3CAf4fE4Cb7;
+    address public constant AMMWBTC = 0x13B2f6928D7204328b0E8E4BCd0379aA06EA21FA;
+    address public constant AMMWETH = 0xf9Fb4AD91812b704Ba883B11d2B576E890a6730A;
```
Comment (0xrajath) [inline src/AMMWithdrawer.sol]: Small nitpick. There are actually `aAMMTokens` instead of `AMMTokens`. Might be better to name it appropriately.

File: src/AMMWithdrawer.sol [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+
+    address public constant DAI = 0x6B175474E89094C44Da98b954EedeAC495271d0F;
+    address public constant USDC = 0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48;
+    address public constant USDT = 0xdAC17F958D2ee523a2206206994597C13D831ec7;
+    address public constant WBTC = 0x2260FAC5E5542a773Aa44fBCfeDf7C193bc2C599;
+    address public constant WETH = 0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2;
+
+    address[5] private ammTokens = [AMMDAI, AMMUSDC, AMMUSDT, AMMWBTC, AMMWETH];
```
Comment (0xrajath) [inline src/AMMWithdrawer.sol]: `aAmmTokens` instead of `ammTokens`

File: src/AMMWithdrawer.sol [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+        for (uint256 i = 0; i < length; ) {
+            address token = ammTokens[i];
+            uint256 amount = IERC20(token).balanceOf(AaveV2Ethereum.COLLECTOR);
+
+            IERC20(token).safeTransferFrom(AaveV2Ethereum.COLLECTOR, address(this), amount);
+
+            AaveV2EthereumAMM.POOL.withdraw(tokens[i], type(uint256).max, AaveV2Ethereum.COLLECTOR);
+            unchecked {
+                ++i;
+            }
```
Comment (0xrajath) [inline src/AMMWithdrawer.sol]: No major opinion here. But is it really worthwhile to have this is as unchecked instead of just having it in the for loop for a bit of gas saving? I can see how this will in no way overflow. But IMO it's more readable if it's just in the for loop. 

Upto you though. No major opinion as mentioned earlier

File: src/AMMWithdrawer.sol [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+    address public constant USDC = 0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48;
+    address public constant USDT = 0xdAC17F958D2ee523a2206206994597C13D831ec7;
+    address public constant WBTC = 0x2260FAC5E5542a773Aa44fBCfeDf7C193bc2C599;
+    address public constant WETH = 0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2;
+
+    address[5] private ammTokens = [AMMDAI, AMMUSDC, AMMUSDT, AMMWBTC, AMMWETH];
+    address[5] private tokens = [DAI, USDC, USDT, WBTC, WETH];
+
+    /// Withdraw AMM tokens from Aave V2 Collector Contract
+    function redeem() external {
```
Comment (0xrajath) [inline src/AMMWithdrawer.sol:14]: I have a feeling BGD is going to ask you to add an additional step to deposit the redeemed asset into the regular pool so that it earns yield instead of just sitting idle. Nothing to do here as of now since you've implemented exactly what's written in the governance forum. Just giving you a heads up.

File: src/AaveV2CollectorContractConsolidation.sol [hunk 14]
```diff
@@ -0,0 +1,258 @@
...
+
+        USDC_DECIMALS = USDC.decimals();
+        ETH_USD_ORACLE_DECIMALS = AggregatorV3Interface(ETH_USD_FEED).decimals();
+    }
+
+    /// @notice Swaps USDC for specified token
+    /// @param _token the address of the token to swap USDC for
+    /// @param _amountOut the amount of token wanted
+    /// @dev User has to approve USDC transfer prior to calling swap
+    function swap(address _token, uint256 _amountOut) external {
```
Comment (0xrajath) [inline src/AaveV2CollectorContractConsolidation.sol]: Small nitpick. And no major opinion here. I think this function should be called `purchase` instead of `swap`. It's more reflective of what's actually happening in this function logic.

File: src/AaveV2CollectorContractConsolidation.sol [hunk 15]
```diff
@@ -0,0 +1,258 @@
...
+
+    mapping(address => Asset) public assets;
+
+    constructor() {
+        assets[ARAI] = Asset(
+            ERC20(ARAI).balanceOf(AaveV2Ethereum.COLLECTOR),
+            100,
+            0x483d36F6a1d063d580c7a24F9A42B346f3a69fbb,
+            ERC20(ARAI).decimals(),
+            8,
```
Comment (0xrajath) [inline src/AaveV2CollectorContractConsolidation.sol]: Since you're anyway doing this through the constructor and saving it. IMO, I think it makes sense to fetch the decimals directly from the Chainlink feed here instead of having it hardcoded for all these assets. It's easier to reason about that way and anyone looking at this code can be assured that it's getting the right decimals instead of having to cross-check it separately.

File: src/ProposalPayload.sol [hunk 16]
```diff
@@ -1,17 +1,193 @@
...
+    address public constant AENS = 0x9a14e23A58edf4EFDcB360f68cd1b95ce2081a2F;
+    address public constant ADPI = 0x6F634c6135D2EBD550000ac92F494F9CB8183dAe;
+
+    // AMM Tokens
+
+    address public constant AMMDAI = 0x79bE75FFC64DD58e66787E4Eae470c8a1FD08ba4;
+    address public constant AMMUSDC = 0xd24946147829DEaA935bE2aD85A3291dbf109c80;
+    address public constant AMMUSDT = 0x17a79792Fe6fE5C95dFE95Fe3fCEE3CAf4fE4Cb7;
+    address public constant AMMWBTC = 0x13B2f6928D7204328b0E8E4BCd0379aA06EA21FA;
+    address public constant AMMWETH = 0xf9Fb4AD91812b704Ba883B11d2B576E890a6730A;
```
Comment (0xrajath) [inline src/ProposalPayload.sol]: `AAMMTokens` instead of just `AMMTokens`

File: src/AMMWithdrawer.sol [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+        uint256 length = ammTokens.length;
+
+        for (uint256 i = 0; i < length; ) {
+            address token = ammTokens[i];
+            uint256 amount = IERC20(token).balanceOf(AaveV2Ethereum.COLLECTOR);
+
+            IERC20(token).safeTransferFrom(AaveV2Ethereum.COLLECTOR, address(this), amount);
+
+            AaveV2EthereumAMM.POOL.withdraw(tokens[i], type(uint256).max, AaveV2Ethereum.COLLECTOR);
```
Comment (0xrajath) [inline src/AMMWithdrawer.sol:27]: Just realized something. And maybe for you to check. What if there is no aAMMToken to redeem here for one of the assets (i.e the balance is 0)? Does withdraw revert in that case? If yes, then in this case, even if one asset as a 0 balance, then the entire redeem might not work for all the assets.

Comment (0xrajath) [reply src/AMMWithdrawer.sol:27]: This is more applicable to `redeem()` being called later down the line rather than the first time in the proposal payload.

File: src/AaveV2CollectorContractConsolidation.sol [hunk 18]
```diff
@@ -0,0 +1,258 @@
...
+import {SafeERC20} from "@openzeppelin/token/ERC20/utils/SafeERC20.sol";
+
+/// @title AaveV2CollectorContractConsolidation
+/// @author Llama
+/// @notice Contract to swap excess assets for USDC at a discount
+contract AaveV2CollectorContractConsolidation {
+    using SafeERC20 for ERC20;
+
+    uint256 public immutable USDC_DECIMALS;
+    uint256 public immutable ETH_USD_ORACLE_DECIMALS;
```
Comment (0xrajath) [inline src/AaveV2CollectorContractConsolidation.sol:17]: Why not just have these 2 as constants instead of setting them as immutable and instantiating through the construcutor? They're static values right?

File: src/AaveV2CollectorContractConsolidation.sol [hunk 19]
```diff
@@ -0,0 +1,258 @@
...
+        );
+
+        USDC_DECIMALS = USDC.decimals();
+        ETH_USD_ORACLE_DECIMALS = AggregatorV3Interface(ETH_USD_FEED).decimals();
+    }
+
+    /// @notice Swaps USDC for specified token
+    /// @param _token the address of the token to swap USDC for
+    /// @param _amountOut the amount of token wanted
+    /// @dev User has to approve USDC transfer prior to calling swap
```
Comment (0xrajath) [inline src/AaveV2CollectorContractConsolidation.sol]: You can add an additional @dev natspec here stating that they can use `type(uint256).max` to get the entire amount.

File: src/AaveV2CollectorContractConsolidation.sol [hunk 20]
```diff
@@ -0,0 +1,258 @@
...
+        if (asset.ethFeedOnly) {
+            uint256 ethUsdPrice = getOraclePrice(ETH_USD_FEED);
+            oraclePrice *= ethUsdPrice;
+            exponent += ETH_USD_ORACLE_DECIMALS;
+        }
+
+        // Basis points arbitrage incentive
+        amountIn =
+            (((_amountOut * oraclePrice) / 10**exponent) * 10000) / // Amount before discount
+            (10000 + asset.premium);
```
Comment (0xrajath) [inline src/AaveV2CollectorContractConsolidation.sol]: Would recommend doing all the multiplications first followed by all the divisions so that no precision is lost here. I would also highly recommend adding in comments before this statement to show the calculation exactly so that it's clear to anyone looking at the code to know what exactly is going on here.

File: src/test/ProposalPayloadE2E.t.sol [hunk 21]
```diff
@@ -1,39 +1,238 @@
...
 
-    function testExecute() public {
-        // Pre-execution assertations
+    function testWithdrawOfAMMTokens() public {
+        GovHelpers.passVoteAndExecute(vm, proposalId);
+
+        uint256 length = ammTokens.length;
+        for (uint256 i = 0; i < length; i++) {
+            // The withdrawal conversion is not 1 to 1 so might not be able to redeem to 0
+            assertApproxEqAbs(IERC20(ammTokens[i]).balanceOf(AaveV2Ethereum.COLLECTOR), 0, 4 ether);
```
Comment (0xrajath) [inline src/test/ProposalPayloadE2E.t.sol]: Curious as to the 4 ether approximation here. Are there cases where that much is left behind?

Comment (0xrajath) [reply src/test/ProposalPayloadE2E.t.sol]: Can you also add an additional assertion for the underlying asset that's received? And also another assertion to check that no asset is left behind in the AMM Withdrawer either (since you're transferring the aAMMToken into that contract)

File: src/test/ProposalPayloadE2E.t.sol [hunk 22]
```diff
@@ -1,39 +1,238 @@
...
+        }
+    }
+
+    function testAllowanceOfToken() public {
```
Comment (0xrajath) [inline src/test/ProposalPayloadE2E.t.sol:189]: Can you add a loop inside this to check for all tokens and not just ARAI?

File: src/AaveV2CollectorContractConsolidation.sol [hunk 23]
```diff
@@ -0,0 +1,258 @@
...
+
+import {AaveV2Ethereum} from "@aave-address-book/AaveV2Ethereum.sol";
+import {AggregatorV3Interface} from "./external/AggregatorV3Interface.sol";
+import {ERC20} from "@openzeppelin/token/ERC20/ERC20.sol";
+import {SafeERC20} from "@openzeppelin/token/ERC20/utils/SafeERC20.sol";
+
+/// @title AaveV2CollectorContractConsolidation
+/// @author Llama
+/// @notice Contract to swap excess assets for USDC at a discount
+contract AaveV2CollectorContractConsolidation {
```
Comment (0xrajath) [inline src/AaveV2CollectorContractConsolidation.sol:13]: You might need to add a rescue function as well for tokens that are accidentally sent directly to this contract. I'm pretty sure BGD is going to ask for that. You can take a look at the 100K BAL Bonding curve contract on how to do that: https://github.com/llama-community/aave-bal-bonding-curve/blob/main/src/OneWayBondingCurve.sol#L116-L122

And ofc, add appropriate test cases for that.

**Review by 0xrajath — APPROVED** [review]
LGTM. Left some comments

File: src/AMMWithdrawer.sol (hunk 13 above)
Comment (efecarranza) [reply src/AMMWithdrawer.sol:14]: That makes sense, I'll look for that on their review

File: src/AMMWithdrawer.sol [hunk 24]
```diff
@@ -0,0 +1,48 @@
...
+    function redeem() external {
+        uint256 length = aAmmTokens.length;
+
+        for (uint256 i = 0; i < length; ) {
+            address token = aAmmTokens[i];
+            uint256 amount = IERC20(token).balanceOf(AaveV2Ethereum.COLLECTOR);
+
+            if (amount == 0) {
+                continue;
+            }
```
Comment (0xrajath) [inline src/AMMWithdrawer.sol:25]: Can we have a test to check this specific condition?

File: src/AaveV2CollectorContractConsolidation.sol [hunk 25]
```diff
@@ -0,0 +1,275 @@
...
+            =>     amountInWithDiscount = amountIn * 10000 / (10000 + 300) 
+         */
+        amountIn =
+            (((_amountOut * oraclePrice) / 10**exponent) * 10000) / // Amount before discount
+            (10000 + asset.premium);
+    }
+
+    /// @return The oracle price
+    /// @notice The peg price of the referenced oracle as USD per unit
+    function getOraclePrice(address _feedAddress) public view returns (uint256) {
```
Comment (0xrajath) [inline src/AaveV2CollectorContractConsolidation.sol:262]: Curious on your thoughts here. Do you think it would be better to pass in the token address rather than the feed address into this function from a UI/UX perspective?

File: src/AaveV2CollectorContractConsolidation.sol [hunk 26]
```diff
@@ -0,0 +1,275 @@
...
+            Amount In Calculation with 3% discount (300 basis points)
+            The actual calculation is a collapsed version of this to prevent precision loss:
+            => amountIn = (amountTokenWei / 10^tokenDecimals) * (chainlinkPrice / chainlinkPrecision) * 10^usdcDecimals
+            => amountInWithDiscount = amountIn * 10000 / (10000 + bps)
+            => ie: amountIn = (amountTokenWei / 10^18) * (chainlinkPrice / 10^8) * 10^6
+            =>     amountInWithDiscount = amountIn * 10000 / (10000 + 300) 
+         */
+        amountIn =
+            (((_amountOut * oraclePrice) / 10**exponent) * 10000) / // Amount before discount
+            (10000 + asset.premium);
```
Comment (0xrajath) [inline src/AaveV2CollectorContractConsolidation.sol]: How I think this should look like (Let me know what you think):

```
/** 
    The actual calculation is a collapsed version of this to prevent precision loss:
    => amountIn = (amountTokenWei / 10^tokenDecimals) * (chainlinkPrice / chainlinkPrecision) * 10^usdcDecimals
    => amountInWithDiscount = amountIn * 10000 / (10000 + premium)
   
    Example for asset with 18 decimals, chainlink precision of 8 decimals and 3% premium on asset being purchased (300 bps)
    => ie: amountIn = (amountTokenWei / 10^18) * (chainlinkPrice / 10^8) * 10^6
    =>      amountInWithDiscount = amountIn * 10000 / (10000 + 300)  
*/
amountIn = (_amountOut * oraclePrice * 10000) / (10**exponent * (10000 + asset.premium))
```

File: src/AaveV2CollectorContractConsolidation.sol (hunk 25 above)
Comment (efecarranza) [reply src/AaveV2CollectorContractConsolidation.sol:262]: I wanted to make it simpler and they can get the oracle address from reading the contract so I went with this approach. I don't expect it'll have a front end component that would make this harder to use but I can be convinced to refactor, I'm ok either way.

File: src/AaveV2CollectorContractConsolidation.sol (hunk 26 above)
Comment (0xrajath) [reply src/AaveV2CollectorContractConsolidation.sol]: @efecarranza Can you adjust the comment to also look like the above? I think that makes it very clear as to what's happening.

**Review by 0xrajath — APPROVED** [review]

File: src/AaveV2CollectorContractConsolidation.sol [hunk 27]
```diff
@@ -0,0 +1,294 @@
...
+/// @title AaveV2CollectorContractConsolidation
+/// @author Llama
+/// @notice Contract to sell excess assets for USDC at a discount
+contract AaveV2CollectorContractConsolidation {
+    using SafeERC20 for ERC20;
+
+    uint256 public immutable USDC_DECIMALS;
+    uint256 public immutable ETH_USD_ORACLE_DECIMALS;
+
+    address public constant ARAI = TokenAddresses.ARAI;
```
Comment (sendra) [inline src/AaveV2CollectorContractConsolidation.sol]: Would remove constants, as they are no longer used

File: src/AaveV2CollectorContractConsolidation.sol [hunk 28]
```diff
@@ -0,0 +1,294 @@
...
+    address public constant AMANA = TokenAddresses.AMANA;
+    address public constant MANA = TokenAddresses.MANA;
+    address public constant ABUSD = TokenAddresses.ABUSD;
+    address public constant BUSD = TokenAddresses.BUSD;
+    address public constant ZRX = TokenAddresses.ZRX;
+    address public constant AZRX = TokenAddresses.AZRX;
+    address public constant AENS = TokenAddresses.AENS;
+    address public constant ADPI = TokenAddresses.ADPI;
+
+    ERC20 public constant USDC = ERC20(0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48);
```
Comment (sendra) [inline src/AaveV2CollectorContractConsolidation.sol]: Would use address from lib

File: src/AaveV2CollectorContractConsolidation.sol [hunk 29]
```diff
@@ -0,0 +1,276 @@
...
+        bool ethFeedOnly;
+    }
+
+    mapping(address => Asset) public assets;
+
+    constructor() {
+        assets[TokenAddresses.ARAI] = Asset(
+            ERC20(TokenAddresses.ARAI).balanceOf(AaveV2Ethereum.COLLECTOR),
+            100,
+            0x483d36F6a1d063d580c7a24F9A42B346f3a69fbb,
```
Comment (sendra) [inline src/AaveV2CollectorContractConsolidation.sol]: The price feed addresses should be set as constants (in lib much better). This way will be much better to understand what they are, and much easier to read / check

File: src/ProposalPayload.sol [hunk 30]
```diff
@@ -1,17 +1,167 @@
...
+    constructor(address _consolidationContract, AMMWithdrawer _withdrawContract) {
+        consolidationContract = _consolidationContract;
+        withdrawContract = _withdrawContract;
+    }
+
     /// @notice The AAVE governance executor calls this function to implement the proposal.
-    function execute() external {}
+    function execute() external {
+        // Transfer to withdraw contract to spend pre-defined amount of tokens and then redeem AMM Tokens
+        AaveMisc.AAVE_ECOSYSTEM_RESERVE_CONTROLLER.transfer(
```
Comment (sendra) [inline src/ProposalPayload.sol]: The transfers could be done with a loop, using the lib getAMMTokens method.

File: src/ProposalPayload.sol [hunk 31]
```diff
@@ -1,17 +1,167 @@
...
+            AaveV2Ethereum.COLLECTOR,
+            TokenAddresses.aAMMWETH,
+            address(withdrawContract),
+            IERC20(TokenAddresses.aAMMWETH).balanceOf(AaveV2Ethereum.COLLECTOR)
+        );
+
+        withdrawContract.redeem();
+
+        // Approve the Consolidation Contract to spend pre-defined amount of tokens from AAVE V2 Collector
+        AaveMisc.AAVE_ECOSYSTEM_RESERVE_CONTROLLER.approve(
```
Comment (sendra) [inline src/ProposalPayload.sol]: This could also be made into a loop by creating a method on the lib to get all the necessary tokens

File: src/AaveV2CollectorContractConsolidation.sol [hunk 32]
```diff
@@ -0,0 +1,276 @@
...
+        uint8 decimals;
+        uint8 oracleDecimals;
+        bool ethFeedOnly;
+    }
+
+    mapping(address => Asset) public assets;
+
+    constructor() {
+        assets[TokenAddresses.ARAI] = Asset(
+            ERC20(TokenAddresses.ARAI).balanceOf(AaveV2Ethereum.COLLECTOR),
```
Comment (sendra) [inline src/AaveV2CollectorContractConsolidation.sol:46]: Using the balanceOf is ok
